File: 08/solution.py
import pathlib

import numpy as np


def left_only(input):
    n, _ = input.shape
    left = np.zeros(input.shape) - 1
    for i in range(1, n):
        left[:, i] = np.maximum(input[:, i - 1], left[:, i - 1])
    return left


def part1(input):
    left = left_only(input)
    right = np.fliplr(left_only(np.fliplr(input)))
    top = left_only(input.T).T
    bottom = np.fliplr(left_only(np.fliplr(input.T))).T

    minv = np.min([left, right, top, bottom], axis=0)
    return np.sum(minv < input)


def part2(input):
    def helper(inp):
        n, m = inp.shape
        view = np.zeros(inp.shape, dtype=int)
        for i in range(n - 1):
            for j in range(m):
                base = inp[i, j]
                s = 1
                while s + i < n - 1 and inp[s + i, j] < base:
                    s += 1
                view[i, j] = s
        return view

    down = helper(input)
    up = np.flipud(helper(np.flipud(input)))
    right = helper(input.T).T
    left = np.flipud(helper(np.flipud(input.T))).T

    return np.max(np.prod([down, up, left, right], axis=0))


def parse(path):
    data = []
    with open(path) as f:
        for line in f.readlines():
            data.append([int(v) for v in line.strip()])
    return np.array(data)


def main():
    input_file = parse("input.dat")
    test_input_file = parse("test_input.dat")

    assert part1(test_input_file) == 21
    print(part1(input_file))

    assert part2(test_input_file) == 8
    print(part2(input_file))


if __name__ == "__main__":
    main()

File: 08/solution.hs
import Data.Char
import Data.List

parseInput :: String -> [[Int]]
parseInput s = map (map digitToInt) . lines $ s

fromLeft :: [[Int]] -> [[Int]]
fromLeft d = res where
  tmp = map (scanl max (-1)) d
  res = map (reverse . tail . reverse) tmp

part1 :: [[Int]] -> Int
part1 d = res where
  left = fromLeft d
  right = map reverse (fromLeft (map reverse d))
  top = transpose (fromLeft (transpose d))
  bottom_d = map reverse (transpose d)
  bottom = transpose (map reverse (fromLeft bottom_d))

  helper1d :: [Int] -> [Int] -> [Int]
  helper1d l r = (map (\t -> min (fst t) (snd t)) (zip l r))

  helper2d :: [[Int]] -> [[Int]] -> [[Int]]
  helper2d l r = map (\t -> (helper1d (fst t) (snd t))) (zip l r)

  helper1db l d = (map (\t -> if (fst t) < (snd t) then 1 else 0) (zip l d))
  helper2db l d = map (\t -> (helper1db (fst t) (snd t))) (zip l d)

  min_height_to_edge = helper2d (helper2d left right) (helper2d top bottom)
  is_visible = helper2db min_height_to_edge d
  res = sum (map sum is_visible)

main = do
  input <- readFile "input.dat"
  let forrest = parseInput input
  -- print (fromLeft forrest)
  print (part1 forrest)
